harp/reader.py

```diff
@@ -44,13 +45,34 @@ def __init__(
         self.read = read
 
 
+class RegisterMap(UserDict[str, RegisterReader]):
+    _address_map: dict[int, RegisterReader]
+
+    def __init__(self, registers: dict[str, RegisterReader]) -> None:
+        super().__init__(registers)
+        self._address_map = {
+            value.register.address: value for value in registers.values()
+        }
+
+    def __getitem__(self, __key: Union[str, int]) -> RegisterReader:
+        if isinstance(__key, int):
+            return self._address_map[__key]
+        else:
+            return super().__getitem__(__key)
+
+
 class DeviceReader:
     device: Model
-    registers: dict[str, RegisterReader]
+    registers: RegisterMap
 
-    def __init__(self, device: Model, registers: dict[str, RegisterReader]) -> None:
+    def __init__(
+        self, device: Model, registers: Union[RegisterMap, dict[str, RegisterReader]]
+    ) -> None:
         self.device = device
-        self.registers = registers
+        if isinstance(registers, RegisterMap):
+            self.registers = registers
+        else:
+            self.registers = RegisterMap(registers)
```

Comment:
Basically even if registers is an instance of `RegisterMap` it still behaves as a valid dict and in this case it will be nice to enforce a new copy anyway for consistency.

```suggestion
        self.registers = RegisterMap(registers)
```